models/policies/sac.py

```diff
@@ -0,0 +1,283 @@
+# Code adopted from https://github.com/pranz24/pytorch-soft-actor-critic
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.optim import Adam
+from torch.distributions import Normal, kl_divergence
+
+from models.network import TaskPolicy, QNetwork
+
+
+def soft_update(target, source, tau):
+    for target_param, param in zip(target.parameters(), source.parameters()):
+        target_param.data.copy_(target_param.data * (1.0 - tau) + param.data * tau)
+
+
+def hard_update(target, source):
+    target.load_state_dict(source.state_dict())
+
+
+class SAC(nn.Module):
+    def __init__(
+        self,
+        dim_s,
+        dim_a,
+        hidden_size=256,
+        num_layers=3,
+        gamma=0.99,
+        tau=1e-2,
+        alpha=0.2,
+        alpha_min=0.,
+        alpha_max=1e6,
+        target_update_interval=1,
+        automatic_entropy_tuning=True,
+        max_entropy_range=100.,
+        kld_penalty_coeff=0.,
+        kld_max=100.,
+        action_tanh=False,
+        action_min=-1,
+        action_max=1,
+        adjust_action_range=False,
+    ):
+        super().__init__()
+        self.gamma = gamma
+        self.tau = tau
+        self.val_alpha = alpha
+        self.alpha_min = alpha_min
+        self.alpha_max = alpha_max
+        self.target_update_interval = target_update_interval
+        self.automatic_entropy_tuning = automatic_entropy_tuning
+        self.max_entropy_range = max_entropy_range
+        self.kld_penalty_coeff = kld_penalty_coeff
+        self.kld_max = kld_max
+        self.action_tanh = action_tanh
+        self.adjust_action_range = adjust_action_range
+        self.action_min = torch.as_tensor(
```

Comment:
I have trouble understanding the `action_min` and `action_max` code. Could you add some comments about it in the code?

Reply:
Sure. Maybe I should separate them to a function.